contract_compassion/model/contracts.py

```diff
@@ -0,0 +1,646 @@
+
+    def on_change_group_id(self, cr, uid, ids, group_id, context=None):
+        """ Compute next invoice_date """
+        res = dict()
+        current_date = datetime.today()
+        is_active = False
+        if ids:
+            contract = self.browse(cr, uid, ids[0], context)
+            if contract.state not in (
+                    'draft', 'mandate') and contract.next_invoice_date:
+                is_active = True
+                current_date = datetime.strptime(contract.next_invoice_date,
+                                                 DF)
+        if group_id:
+            contract_group = self.pool.get('recurring.contract.group').browse(
+                cr, uid, group_id, context)
+            if contract_group.next_invoice_date:
+                next_group_date = datetime.strptime(
+                    contract_group.next_invoice_date, DF)
+                next_invoice_date = current_date.replace(
+                    day=next_group_date.day)
+            else:
+                next_invoice_date = current_date.replace(day=1)
+            payment_term = contract_group.payment_term_id.name
+        else:
+            next_invoice_date = current_date.replace(day=1)
+            payment_term = ''
+
+        if current_date.day > 15 or (
+                payment_term in ('LSV', 'Postfinance') and not is_active):
+            next_invoice_date = next_invoice_date + relativedelta(months=+1)
+        res['value'] = {'next_invoice_date': next_invoice_date.strftime(DF)}
+        return res
+
+    def contract_waiting(self, cr, uid, ids, context=None):
+        for contract in self.browse(cr, uid, ids, {'lang': 'en_US'}):
+            payment_term = contract.group_id.payment_term_id.name
+            if 'LSV' in payment_term or 'Postfinance' in payment_term:
+                # Recompute next_invoice_date
+                today = datetime.today()
+                next_invoice_date = datetime.strptime(
+                    contract.next_invoice_date, DF).replace(month=today.month,
+                                                            year=today.year)
+                if today.day > 15 and next_invoice_date.day < 15:
+                    next_invoice_date = next_invoice_date + relativedelta(
+                        months=+1)
+                contract.write({
+                    'next_invoice_date': next_invoice_date.strftime(DF)})
+
+        self.write(cr, uid, ids, {'state': 'waiting'}, context)
+        return True
+
+    def contract_cancelled(self, cr, uid, ids, context=None):
+        self.write(cr, uid, ids, {'state': 'cancelled'}, context)
+        # Remove the sponsor of the child
+        for contract in self.browse(cr, uid, ids, context):
+            if contract.child_id:
+                contract.child_id.write({'sponsor_id': False})
+        return True
+
+    def contract_terminated(self, cr, uid, ids, context=None):
+        self.write(cr, uid, ids, {'state': 'terminated'})
+
+        ctx = {'lang': 'en_US'}

```

Comment:
All this part up to the end of the procedure can be moved to sponsorship_compassion. Also this should concern only sponsorships contracts.